lib/AST/Availability.cpp

```diff
@@ -321,11 +321,29 @@ llvm::Optional<AvailableAttrDeclPair> Decl::getSemanticUnavailableAttr() const {
                            llvm::None);
 }
 
-static bool isUnconditionallyUnavailable(const Decl *D) {
-  if (auto unavailableAttrAndDecl = D->getSemanticUnavailableAttr())
-    return unavailableAttrAndDecl->first->isUnconditionallyUnavailable();
+static bool shouldStubOrSkipUnavailableDecl(const Decl *D) {
+  // Don't trust unavailability on declarations from clang modules.
+  if (isa<ClangModuleUnit>(D->getDeclContext()->getModuleScopeContext()))
+    return false;
 
-  return false;
+  auto unavailableAttrAndDecl = D->getSemanticUnavailableAttr();
+  if (!unavailableAttrAndDecl)
+    return false;
+
+  // getSemanticUnavailableAttr() can return an @available attribute that makes
+  // its declaration unavailable conditionally due to deployment target. Only
+  // stub or skip a declaration that is unavailable regardless of deployment
+  // target.
+  auto *unavailableAttr = unavailableAttrAndDecl->first;
+  if (!unavailableAttr->isUnconditionallyUnavailable())
+    return false;
+
+  // If the decl is only unavailable to app extensions it still may need to be
+  // present at runtime for non-extension processes.
+  if (isApplicationExtensionPlatform(unavailableAttr->Platform))
+    return false;
```

Comment:
When this happens, should we look for a non-app extension availability attribute instead of just returning `false`? Think of a test case like:

```swift
@available(macOS, unavailable)
@available(macOSApplicationExtension, unavailable)
public func doublyUnavailableOnMacOSFunc() {}
```

(If you want to do that, a better approach might be to make `getSemanticUnavailableAttr()` take a flag to ignore app extension platforms.)

Reply:
I considered that, but I really think we shouldn't accept the code in your example. Having both attributes is redundant, and rather than complicating querying I think the AST should be guaranteed to not have this situation.

Reply:
*Do* we reject the code, though? And are there no situations—for instance, when we inherit availability attributes from another platform—where this can happen in a more reasonable example?

```
// Which one of these do we even use if you're building for Mac Catalyst with -app-extension???
@available(macCatalyst, unavailable)
@available(iOSApplicationExtension, unavailable)
public func confusingDiamondAvailabilityInheritanceFunc() {}
```

Reply:
Ok, it took a fair bit of plumbing of new parameters but I've handled this case.